gensim/models/keyedvectors.py

```diff
@@ -1696,8 +1696,8 @@ def intersect_word2vec_format(self, fname, lockf=0.0, binary=False, encoding='ut
             msg=f"merged {overlap_count} vectors into {self.vectors.shape} matrix from {fname}",
         )
 
-    def vectors_for_all(self, keys: Iterable) -> 'KeyedVectors':
-        """Produces vectors for all keys in a given iterable.
+    def vectors_for_all(self, keys: Union[Iterable, Dictionary]) -> 'KeyedVectors':
```

Comment:
Why do we single out `Dictionary`?

Accepting only a sequence of strings seems cleaner API. Ppl who have `Dictionary` can simply pass `Dictionary.token2id`, no?

Reply:
I see what you mean, but am -1 on complicating the API with two input types. IMO that increases the cognitive load and makes people look up the docs, rather than the opposite.

Where one code path merges into the other via a simple `input = input.some_attribute`, that can (should) be done by the caller. A cleaner / leaner API is preferable.

Reply:
That's not what I see in this PR though:

```python
vocabulary = [key for key in keys.token2id if key in self]
```

Reply:
Indeed, and it's likely not even what we want. As discussed with @gojomo in https://github.com/RaRe-Technologies/gensim/pull/3157#issuecomment-848081847, we likely want to order the dictionary by decreasing frequency for cache locality. Nevertheless, this is a good reason to special-case `Dictionary`, since it's not easy to get this right.

Reply:
I will convert this to draft, since there are pending changes.

Reply:
Is this solution OK for you?